src/lib/hydrate.ts

```diff
@@ -186,8 +187,18 @@ export const hydrate =
           // the matching marker.
           currentNodePart.startNode = node;
 
-          // Check for a template result digest
-          if (value instanceof TemplateResult) {
+          // The following logic closely follows the NodePart commit() cascade
```

Comment:
This function is crazy long now. Seems like it could benefit from some factoring. Perhaps we could start with this commit cascade?

On that function, I'd love to see a comment that calls out the ordering:
1. directive
2. noChange
3. primitive
4. TemplateResult
5. Node
6. Iterable
7. nothing
8. Everything else

Reply:
I factored `hydrate` into:
* `openNodePart()`
* `createAttributeParts()`
* `closeNodePart()`

which are called from a much tighter while loop now in `hydrate`'s treewalk over the comments, which I think is a nice improvement.  

I tried several ways to separate the commit cascade out from the other work of `openNodePart`, but it was pretty unsatisfying.

I added the order comment for review, although I don't personally think this helps